### DISCUSSION TITLE
Dockerfile Custom Providers Example

### DISCUSSION THREAD

**Discussion by samdammers** [keycloak/keycloak#10756] [Feedback]
@pedroigor  Just going through an upgrade from 15->17.

When using a single Dockerfile, It didn't take me long to figure out I needed to drop them in twice (unless I am doing something wrong?), Perhaps something to mention in the Providers guide for the 18 release? https://github.com/keycloak/keycloak/pull/10401

Not copying them back in results in

```
ERROR: Failed to start server in (production) mode
ERROR: Failed to open /opt/keycloak/lib/../providers/extensions-17.0.0.jar
ERROR: /opt/keycloak/lib/../providers/extensions-17.0.0.jar
```

Cutdown example

```
FROM quay.io/keycloak/keycloak:17.0.0 as builder

# Copy in customised extensions for build step (optimises classpath)
COPY ./extensions-17.0.0.jar /opt/keycloak/providers

RUN /opt/keycloak/bin/kc.sh build

FROM quay.io/keycloak/keycloak:17.0.0
COPY --from=builder /opt/keycloak/lib/quarkus/ /opt/keycloak/lib/quarkus/
WORKDIR /opt/keycloak

# Copy in customised extensions take 2 (so the jar is present when server started)
COPY ./extensions-17.0.0.jar /opt/keycloak/providers

ENTRYPOINT ["/opt/keycloak/bin/kc.sh", "start"]
```

**Comment by sergej-reiser** [comment]
I stumbled upon the same issue. Apparently you also need to copy your providers into your final image. It's not enough to copy them into the builder image. 

A similar question has been asked in  #10502
